src/include/einsums/TensorAlgebraThreads.hpp

```diff
@@ -0,0 +1,1206 @@
+#pragma once
+
+#include "einsums/_Common.hpp"
+#include "einsums/_Compiler.hpp"
+#include "einsums/_Index.hpp"
+#include "einsums/_TensorAlgebraUtilities.hpp"
+
+#include "einsums/LinearAlgebra.hpp"
+#include "einsums/OpenMP.h"
+#include "einsums/Print.hpp"
+#include "einsums/STL.hpp"
+#include "einsums/Section.hpp"
+#include "einsums/Tensor.hpp"
+#include "einsums/Timer.hpp"
+#include "einsums/utility/SmartPointerTraits.hpp"
+
+#include <algorithm>
+#include <atomic>
+#include <cmath>
+#include <complex>
+#include <cstddef>
+#include <functional>
+#include <numeric>
+#include <sstream>
+#include <stdexcept>
+#include <string>
+#include <thread>
+#include <tuple>
+#include <type_traits>
+#include <utility>
+
+#if defined(EINSUMS_USE_CATCH2)
+#    include <catch2/catch_all.hpp>
+#endif
+
+BEGIN_EINSUMS_NAMESPACE_HPP(einsums::tensor_algebra::threads)
+
+namespace detail {
+
+template <typename... CUniqueIndices, typename... AUniqueIndices, typename... BUniqueIndices, typename... LinkUniqueIndices,
+          typename... CIndices, typename... AIndices, typename... BIndices, typename... TargetDims, typename... LinkDims,
+          typename... TargetPositionInC, typename... LinkPositionInLink, template <typename, size_t> typename CType, typename CDataType,
+          size_t CRank, template <typename, size_t> typename AType, typename ADataType, size_t ARank,
+          template <typename, size_t> typename BType, typename BDataType, size_t BRank>
+void einsum_generic_algorithm(const std::tuple<CUniqueIndices...> &C_unique, const std::tuple<AUniqueIndices...> & /*A_unique*/,
+                              const std::tuple<BUniqueIndices...> & /*B_unique*/, const std::tuple<LinkUniqueIndices...> &link_unique,
+                              const std::tuple<CIndices...> & /*C_indices*/, const std::tuple<AIndices...> & /*A_indices*/,
+                              const std::tuple<BIndices...> & /*B_indices*/, const std::tuple<TargetDims...> &target_dims,
+                              const std::tuple<LinkDims...> &link_dims, const std::tuple<TargetPositionInC...> &target_position_in_C,
+                              const std::tuple<LinkPositionInLink...> &link_position_in_link, const CDataType C_prefactor,
+                              CType<CDataType, CRank>                                                                *C,
+                              const std::conditional_t<(sizeof(ADataType) > sizeof(BDataType)), ADataType, BDataType> AB_prefactor,
+                              const AType<ADataType, ARank> &A, const BType<BDataType, BRank> &B, int num_threads, int thread_id,
+                              std::conditional_t<(sizeof(ADataType) > sizeof(BDataType)), ADataType, BDataType> *work,
+                              std::atomic_int                                                                   &synch) {
+    if (thread_id == 0) {
+        LabeledSection0();
```

Comment:
This actually won't time the function. LabeledSection0() is a macro that creates a Section object. Once the Section object goes out of scope the timer is stopped.

The timer::push and timer::pop should be thread safe but are only coded up for OMP threads. I'm not sure what omp_get_thread_num() would return in this instance.

Reply:
Thanks for pointing this out. I didn't actually check what these did, I just copied and pasted. For now, I'll delete them.